### ISSUE TITLE
apply() does set resourceVersion in metadata: kubectl.kubernetes.io/last-applied-configuration

### ISSUE THREAD

**Issue by maikebertpsc** [openshift/jenkins-client-plugin#128]
When openshift.apply(modifiedObjectConfiguration) is used, then the live configuration has metadata: kubectl.kubernetes.io/last-applied-configuration: resourceVersion set.
This will later on result in problems when "oc apply" is used for declarative object management.

For example:

Create a build config using
`oc apply -f bc-app.json`
This will result in a live configuration having no resourceVersion in metadata: kubectl.kubernetes.io/last-applied-configuration-

Then use this plugin and modify that configuration, e.g. the Git ref:
```
def buildConfigSelector = openshift.selector("bc/app")
def buildConfig = buildConfigSelector.object()
buildConfig.spec.source.git.ref = 'release/1.1'
openshift.apply(buildConfig)
```
This works as expected, but additionally set metadata: kubectl.kubernetes.io/last-applied-configuration: resourceVersion

Then use
`oc apply -f bc-app-modified.json`
to modify any other field.
This will fail with error: 

> "The buildconfigs "app" is invalid: metadata.resourceVersion: Invalid value: 0x0: must be specified ..."

This is caused by the fact that the live configuration last-applied-configuration annotation contains resourceVersion, while the new bc-app-modified.json does not contain resourceVersion.
As a result "oc apply" tries to delete the resourceVersion from the live configuration which is forbidden.

Expected:
apply(objectConfiguration) does not touch metadata: kubectl.kubernetes.io/last-applied-configuration of the live configuration.

**Comment by gabemontero** [comment]
@maikebertpsc what version of `oc` are you using when running from the command line, and what version of `oc` are you running from jenkins.

Please provide output the output from running `oc version` both from you command line and from `sh 'oc version'` from a pipeline.

Most likely this is either a) differing oc versions where an incompatibility exists, or b) an oc bug which should be  reported against openshift/origin

I can tell you that his plugin explicitly does not do anything with that annotation.  This plugin just invokes `oc` under the covers.

If you want to see the precise invocation of `oc` this plugin executes with the `openshift.apply(...)` call, you can turn on verbose logging.  See the instructions at https://github.com/openshift/jenkins-client-plugin#troubleshooting

**Comment by maikebertpsc** [comment]
@gabemontero Thanks for your feedback!

All these calls are in fact done from Jenkins.
The oc version in use is:

> oc version
> oc v3.7.0+7ed6862
> kubernetes v1.7.6+a08f5eeb62
> 
> openshift v3.6.173.0.49
> kubernetes v1.6.1+5115d708d7

**Comment by gabemontero** [comment]
hey @maikebertpsc - I see ... the examples in the description like `oc apply -f bc-app-modified.json` confused me.

I'm wondering if you found a compatability bug between oc 3.7 and master 3.6, in which case we'd need to open a bug against openshift origin

The quickest path would be to run your pipeline with verbose turned on, so we can see all the precise oc invocations.  You provided me that full output, and then I can engage the openshift API team.

**Comment by gabemontero** [comment]
Thanks for the verbose log @dbazhal 

I also talked to our CLI developer for @maikebertpsc claims, and I have some feedback.

1) In fact @maikebertpsc , `oc apply` is **SUPPOSE** to be setting that `metadata: kubectl.kubernetes.io/last-applied-configuration` annotation ... that is how it works.  It is the versioning/state machine

And to reiterate, this plugin is just calling `oc apply` ... it is not explicitly setting the annotation ... the `oc apply` call (which does a passthrough to `kubectl appl`) does it

2) He also notes that the `The buildconfigs "app" is invalid: metadata.resourceVersion: Invalid value: 0x0: must be specified ` is indicative of an invalid field, not anything 

Now, that *COULD* be that the selector.object() translation from json to groovy maps and back to json is doing something odd there.

I'll see if I can construct a sample to mimic what @maikebertpsc has done,

**But it would be helpful if you could echo/print the contents of the selector after you modify it, but before you pass it to openshift.deploy**

3) @dbazhal , In your case, that annotation was already set ... I see it on the result returned from the selector call:

```
	Command> oc --server=https://172.30.0.1:443 --certificate-authority=/var/run/secrets/kubernetes.io/serviceaccount/ca.crt --namespace=project-prod --loglevel=8 --token=XXXXX get deploymentconfig/test1-prev -o=json 
	Status> 0
	StdOut>{
    "apiVersion": "apps.openshift.io/v1",
    "kind": "DeploymentConfig",
    "metadata": {
        "annotations": {
            "kubectl.kubernetes.io/last-applied-configuration": "{\"apiVersion\":\"apps.openshift.io/v1\",\"kind\":\"DeploymentConfig\",\"metadata\":{\"annotations\":{},\"creationTimestamp\":\"2018-04-18T12:41:22Z\",\"generation\":9,\"name\":\"test1-prev\",\"namespace\":\"project-prod\",\"resourceVersion\":\"249843\",\"selfLink\":\"/apis/apps.openshift.io/v1/namespaces/project-prod/deploymentconfigs/test1-prev\",\"uid\":\"ccf0ef43-4305-11e8-a276-d00da98d87e1\"},\"spec\":{\"replicas\":2,\"selector\":{\"name\":\"test1-prev\"},\"strategy\":{\"activeDeadlineSeconds\":21600,\"resources\":{},\"rollingParams\":{\"intervalSeconds\":1,\"maxSurge\":\"25%\",\"maxUnavailable\":\"25%\",\"timeoutSeconds\":600,\"updatePeriodSeconds\":1},\"type\":\"Rolling\"},\"template\":{\"metadata\":{\"creationTimestamp\":null,\"labels\":{\"name\":\"test1-prev\"},\"name\":\"test1-prev\"},\"spec\":{\"containers\":[{\"image\":\"docker-registry.default.svc:5000/project-infra/test1@sha256:627c9a01277f0784ed62420cbf135a78d359c49dabc8ddad891ce7ce641ffd7a\",\"imagePullPolicy\":\"Always\",\"name\":\"test1-prev\",\"ports\":[{\"containerPort\":8080,\"protocol\":\"TCP\"}],\"readinessProbe\":{\"exec\":{\"command\":[\"/bin/bash\",\"-c\",\"test -f /usr/share/app/readiness \\u0026\\u0026 /usr/share/app/readiness\"]},\"failureThreshold\":5,\"initialDelaySeconds\":1,\"periodSeconds\":3,\"successThreshold\":1,\"timeoutSeconds\":1},\"resources\":{},\"terminationMessagePath\":\"/dev/termination-log\",\"terminationMessagePolicy\":\"File\"}],\"dnsPolicy\":\"ClusterFirst\",\"restartPolicy\":\"Always\",\"schedulerName\":\"default-scheduler\",\"securityContext\":{},\"terminationGracePeriodSeconds\":30}},\"test\":false,\"triggers\":[{\"type\":\"ConfigChange\"},{\"imageChangeParams\":{\"containerNames\":[\"test1-prev\"],\"from\":{\"kind\":\"ImageStreamTag\",\"name\":\"test1:prod-prev\",\"namespace\":\"project-infra\"},\"lastTriggeredImage\":\"docker-registry.default.svc:5000/project-infra/test1@sha256:627c9a01277f0784ed62420cbf135a78d359c49dabc8ddad891ce7ce641ffd7a\"},\"type\":\"imageChange\"}]},\"status\":{\"availableReplicas\":2,\"conditions\":[{\"lastTransitionTime\":\"2018-04-19T09:09:19Z\",\"lastUpdateTime\":\"2018-04-19T09:09:19Z\",\"message\":\"Deployment config has minimum availability.\",\"status\":\"True\",\"type\":\"Available\"},{\"lastTransitionTime\":\"2018-04-19T15:34:06Z\",\"lastUpdateTime\":\"2018-04-19T15:34:07Z\",\"message\":\"replication controller \\\"test1-prev-6\\\" successfully rolled out\",\"reason\":\"NewReplicationControllerAvailable\",\"status\":\"True\",\"type\":\"Progressing\"}],\"details\":{\"causes\":[{\"type\":\"Manual\"}],\"message\":\"manual change\"},\"latestVersion\":6,\"observedGeneration\":9,\"readyReplicas\":2,\"replicas\":2,\"unavailableReplicas\":0,\"updatedReplicas\":2}}\n"
        },
```

So in theory it is different from @maikebertpsc 's scenario, assuming it was not already present in the `bc-app.json1 from his `oc apply -f bc-app.json` invocation.

That said, your debug confirmed part of what I wanted to confirm ...this plugin is just passing the obj json as is to `oc apply` ... for example:

```
Command> oc --server=https://172.30.0.1:443 --certificate-authority=/var/run/secrets/kubernetes.io/serviceaccount/ca.crt --namespace=project-prod --loglevel=8 --token=XXXXX apply -f /tmp/apply3964716360864902446.markup -o=name 
```

I actually don't see any errors in the output you provided ... the last entry was getting a 200 http OK on the apply call.

**In addition to providing the precise error you are seeing ..it would be helpful if you could echo/print the contents of the selector after you modify it, but before you pass it to openshift.deploy**

thanks,

**Comment by dbazhal** [comment]
@gabemontero oh sorry, my bad, haven't seen the object is already modified. I'll submit the case with dc source state.
Could you please clarify what do you mean by "echo/print the contents of the selector after you modify it"? Should I just print ```prevDeploy``` object before apply?

**Comment by gabemontero** [comment]
On Fri, Apr 20, 2018 at 1:15 PM, Dmitry Bazhal <notifications@github.com>
wrote:

> @gabemontero <https://github.com/gabemontero> oh sorry, my bad, haven't
> seen the object is already modified. I'll submit the case with dc source
> state.
>
np - among other things, I'll reiterate that the annotation being there is
expected, and is a function of how the oc/kubectl apply
command works ... and show the actual error received ... I saw no error in
the verbose log you posted

> Could you please clarify what do you mean by "echo/print the contents of
> the selector after you modify it"? Should I just print prevDeploy object
> before apply?
>
Yep, print that, and newDeploy as well ... I want to see the contents, to
correlate with any  errors reported; it will also help
me in constructing a repro so I can test locally

thanks

**Comment by gabemontero** [comment]
OK, @maikebertpsc @dbazhal - some results from my attempts at reproducing and existing `oc` and `openshift` bug research:

1) for @maikebertpsc using the latest 3.10 dev version, this pipeline:
```
openshift.withCluster() {
    openshift.withProject() {
        def bcsel = openshift.selector('bc/test')
        def bc = bcsel.object()
        bc.spec.source.git.ref = 'release/1.1'
        openshift.apply(bc)
    }
}
```
works just fine

Doing some research on `oc` and `kubect` apply bugs, I see a few wit some similar aspects, such as https://github.com/kubernetes/kubernetes/issues/25241 (which has the same original error as yours about metadata.resourceVersion having an invalid value) and https://github.com/openshift/origin/issues/14965 ... there were also some 3.7 client to 3.6 server issues that were addressed some fixes after 3.7 shipped.

I think for you, you'll need to line up your `oc` and `openshift` to either both be 3.6, 3.7, or even the new 3.9.

2) for @dbazhal, see my discussion in https://github.com/openshift/jenkins-client-plugin/issues/149 and the resulting PR https://github.com/openshift/jenkins-client-plugin/pull/151 ... for DC's especially, it is prudent to reduce the contents of the apply object to the minimal amount needed.

I'm going to close this out with those recommendations .... reopen if your are still having issues after applying the recommendations I noted.

thanks
